File: Leonard_1994.py
import matplotlib.pyplot as plt
import numpy as np
from decimal import *
getcontext().prec=17

def gfg(p,n,k,a,t,c,s):
    q=1-p
    N=(n*(1-k+(1-q**2)*a))/(1-(1-q**2)*t+n*((1-q**2)*(a+t)-k))
    P=(p*(1-c-s*(1-t)+N*s*(k-a-t)))/(1-s+N*k*s+(1-q**2)*(t*s-c-N*s*(a+t)))
    return N,P
def eqpoint(k,a,t,c,s):
    neq=(t*s-c)/(t*s+a*s)
    peq=1-np.sqrt(1-k/(a+t))
    return neq,peq
tmax=165
p_freq=list()
n_freq=list()
K=0.3
A=0.0
T=1.0
C=0.03
S=0.8
PTEMP=0
NTEMP=0
PINITIAL=0.163
NINITIAL=0.2

###################################
#INITIALIZATION
PTEMP=PINITIAL
NTEMP=NINITIAL
p_freq.append(PINITIAL)
n_freq.append(NINITIAL)

for time in range(0,tmax):
    #NTEMP,PTEMP=gfg(Decimal(PTEMP),Decimal(NTEMP),Decimal(K),Decimal(A),Decimal(T),Decimal(C),Decimal(S))
    NTEMP,PTEMP=gfg(PTEMP,NTEMP,K,A,T,C,S)
    
    p_freq.append(PTEMP)
    n_freq.append(NTEMP)
plt.plot(NINITIAL,PINITIAL,'k.')

NEQ,PEQ=eqpoint(K,A,T,C,S)
plt.plot(NEQ,PEQ,'ko')
plt.plot(n_freq,p_freq,'k')
plt.xlim([0,1])
plt.ylim([0,1])
plt.show()

File: Leonard_1994.m
clear all
close all

%Initial Conditions
P(1) = 0.163; %FIgure 2A
N(1) = 0.2; %Figure 2A
%P(1) = 0.163; %FIgure 2B
%N(1) = 0.18; %Figure 2B


%Frequencies
p_freq=[];
n_freq=[];

%Parameters
k=0.3;
a=0.0;
t=1.0;
c=0.03;
s=0.8;

%Generations
tmax=165; %Figure 2A
%tmax=194; %Figure 2B

%Model
for x = 1:tmax
    q=1-P(x);
    N(x+1) = (N(x).*(1-k+(1-q.^2).*a))./(1-(1-q.^2).*t+N(x).*((1-q.^2).*(a+t)-k));
    P(x+1) = P(x).*(1-c-s.*(1-t)+N(x+1).*s.*(k-a-t))./(1-s+N(x+1).*k.*s+(1-q.^2).*(t.*s-c-N(x+1).*s.*(a+t)));
end

neq = (t.*s - c)./(t.*s+a.*s); 
peq = 1-sqrt(1-k./(a+t));

plot(neq,peq,'ko')
hold on
plot(N,P, 'k')
xlim([0 1]), ylim([0 1])